src/lib/shadow-shim-helper-rs/src/syscall_types.rs

```diff
@@ -0,0 +1,307 @@
+/// Represents a pointer to a virtual address in plugin memory.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+#[repr(C)]
+pub struct PluginPtr {
+    val: usize,
+}
+
+impl PluginPtr {
+    pub fn null() -> Self {
+        0usize.into()
+    }
+
+    pub fn is_null(&self) -> bool {
+        self.val == 0
+    }
+}
+
+impl From<PluginPtr> for usize {
+    fn from(v: PluginPtr) -> usize {
+        v.val
+    }
+}
+
+impl From<usize> for PluginPtr {
+    fn from(v: usize) -> PluginPtr {
+        PluginPtr { val: v }
+    }
+}
+
+impl From<u64> for PluginPtr {
+    fn from(v: u64) -> PluginPtr {
+        PluginPtr {
+            val: v.try_into().unwrap(),
+        }
+    }
+}
+
+impl From<PluginPtr> for u64 {
+    fn from(v: PluginPtr) -> u64 {
+        v.val.try_into().unwrap()
+    }
+}
+
+impl std::fmt::Pointer for PluginPtr {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let ptr = self.val as *const libc::c_void;
+        std::fmt::Pointer::fmt(&ptr, f)
+    }
+}
+
+/// Represents a pointer to a *physical* address in plugin memory.
+#[derive(Copy, Clone, Debug)]
+#[repr(C)]
+pub struct PluginPhysicalPtr {
+    val: usize,
+}
+
+impl From<PluginPhysicalPtr> for usize {
+    fn from(v: PluginPhysicalPtr) -> usize {
+        v.val
+    }
+}
+
+impl From<usize> for PluginPhysicalPtr {
+    fn from(v: usize) -> PluginPhysicalPtr {
+        PluginPhysicalPtr { val: v }
+    }
+}
+
+impl From<u64> for PluginPhysicalPtr {
+    fn from(v: u64) -> PluginPhysicalPtr {
+        PluginPhysicalPtr {
+            val: v.try_into().unwrap(),
+        }
+    }
+}
+
+impl From<PluginPhysicalPtr> for u64 {
+    fn from(v: PluginPhysicalPtr) -> u64 {
+        v.val.try_into().unwrap()
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(C)]
+pub struct SysCallArgs {
+    // SYS_* from sys/syscall.h.
+    // (mostly included from
+    // /usr/include/x86_64-linux-gnu/bits/syscall.h)
+    pub number: libc::c_long,
+    pub args: [SysCallReg; 6],
+}
+
+impl SysCallArgs {
+    pub fn get(&self, i: usize) -> SysCallReg {
+        self.args[i]
+    }
+    pub fn number(&self) -> i64 {
+        self.number
+    }
+}
+
+impl PartialEq for SysCallReg {
+    fn eq(&self, other: &Self) -> bool {
+        unsafe { self.as_u64 == other.as_u64 }
+    }
+}
```

Comment:
nit: move this down to after where `SysCallReg` is defined?

Reply:
done